File: su/src/main/java/org/sqlbuilder/su/objects/TermAttr.java
package org.sqlbuilder.su.objects;

import com.articulate.sigma.Formula;

import org.sqlbuilder.common.NotFoundException;
import org.sqlbuilder.common.Utils;
import org.sqlbuilder.su.Kb;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class TermAttr
{
	public final Character attr;

	private static final Character ISFUNCTION = 'y';

	private static final Character ISMATHFUNCTION = 'm';

	private static final Character ISCOMPARISONOP = '~';

	private static final Character ISLOGICALOP = 'l';

	private static final Character ISQUANTIFIER = 'q';

	private static final Character SUBCLASSOFRELATION = 'R';

	private static final Character SUBCLASSOFFUNCTION = 'F';

	private static final Character SUBCLASSOFPREDICATE = 'P';

	private static final Character SUBCLASSOFATTRIBUTE = 'A';

	private static final Character CHILDOFRELATION = 'r';

	private static final Character CHILDOFFUNCTION = 'f';

	private static final Character CHILDOFPREDICATE = 'p';

	private static final Character CHILDOFATTRIBUTE = 'a';

	// C O N S T R U C T

	private TermAttr(final Character attr)
	{
		this.attr = attr;
	}

	public static Collection<TermAttr> make(final Term sumoTerm, final Kb kb) throws NotFoundException
	{
		String term = sumoTerm.getTerm();

		final List<TermAttr> result = new ArrayList<>();

		if (Formula.isFunction(term))
		{
			result.add(new TermAttr(TermAttr.ISFUNCTION));
		}
		if (Formula.isMathFunction(term))
		{
			result.add(new TermAttr(TermAttr.ISMATHFUNCTION));
		}
		if (Formula.isComparisonOperator(term))
		{
			result.add(new TermAttr(TermAttr.ISCOMPARISONOP));
		}
		if (Formula.isLogicalOperator(term))
		{
			result.add(new TermAttr(TermAttr.ISLOGICALOP));
		}
		if (Formula.isQuantifier(term))
		{
			result.add(new TermAttr(TermAttr.ISQUANTIFIER));
		}

		if (kb.childOf(term, "Relation"))
		{
			result.add(new TermAttr(TermAttr.CHILDOFRELATION));
		}
		if (kb.childOf(term, "Predicate"))
		{
			result.add(new TermAttr(TermAttr.CHILDOFPREDICATE));
		}
		if (kb.childOf(term, "Function"))
		{
			result.add(new TermAttr(TermAttr.CHILDOFFUNCTION));
		}
		if (kb.childOf(term, "Attribute"))
		{
			result.add(new TermAttr(TermAttr.CHILDOFATTRIBUTE));
		}

		if (kb.isSubclass(term, "Relation"))
		{
			result.add(new TermAttr(TermAttr.SUBCLASSOFRELATION));
		}
		if (kb.isSubclass(term, "Predicate"))
		{
			result.add(new TermAttr(TermAttr.SUBCLASSOFPREDICATE));
		}
		if (kb.isSubclass(term, "Function"))
		{
			result.add(new TermAttr(TermAttr.SUBCLASSOFFUNCTION));
		}
		if (kb.isSubclass(term, "Attribute"))
		{
			result.add(new TermAttr(TermAttr.SUBCLASSOFATTRIBUTE));
		}

		if (result.isEmpty())
			throw new NotFoundException(term);

		return result;
	}

	// A C C E S S

	public Character getAttr()
	{
		return attr;
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return attr.toString();
	}

	// I N S E R T

	public String dataRow()
	{
		return String.format("%s", //
				Utils.nullableQuotedEscapedString(attr.toString())); // 1
	}
}

File: su/src/main/java/org/semantikos/su/objects/TermAttr.kt
package org.semantikos.su.objects

import org.sigma.core.Formula
import org.semantikos.common.NotFoundException
import org.semantikos.common.Utils.nullableQuotedEscapedString
import org.semantikos.su.Kb

class TermAttr private constructor(
    val attr: Char,
) {

    // T O S T R I N G

    override fun toString(): String {
        return attr.toString()
    }

    // I N S E R T

    fun dataRow(): String {
        return nullableQuotedEscapedString(attr.toString())
    }

    companion object {

        private const val ISFUNCTION = 'y'

        private const val ISMATHFUNCTION = 'm'

        private const val ISCOMPARISONOP = '~'

        private const val ISLOGICALOP = 'l'

        private const val ISQUANTIFIER = 'q'

        private const val SUBCLASSOFRELATION = 'R'

        private const val SUBCLASSOFFUNCTION = 'F'

        private const val SUBCLASSOFPREDICATE = 'P'

        private const val SUBCLASSOFATTRIBUTE = 'A'

        private const val CHILDOFRELATION = 'r'

        private const val CHILDOFFUNCTION = 'f'

        private const val CHILDOFPREDICATE = 'p'

        private const val CHILDOFATTRIBUTE = 'a'

        @Throws(NotFoundException::class)
        fun make(sumoTerm: Term, kb: Kb): MutableCollection<TermAttr> {
            val term = sumoTerm.term

            val result: MutableList<TermAttr> = ArrayList<TermAttr>()

            if (Formula.isFunction(term)) {
                result.add(TermAttr(ISFUNCTION))
            }
            if (Formula.isMathFunction(term)) {
                result.add(TermAttr(ISMATHFUNCTION))
            }
            if (Formula.isComparisonOperator(term)) {
                result.add(TermAttr(ISCOMPARISONOP))
            }
            if (Formula.isLogicalOperator(term)) {
                result.add(TermAttr(ISLOGICALOP))
            }
            if (Formula.isQuantifier(term)) {
                result.add(TermAttr(ISQUANTIFIER))
            }

            if (kb.childOf(term, "Relation")) {
                result.add(TermAttr(CHILDOFRELATION))
            }
            if (kb.childOf(term, "Predicate")) {
                result.add(TermAttr(CHILDOFPREDICATE))
            }
            if (kb.childOf(term, "Function")) {
                result.add(TermAttr(CHILDOFFUNCTION))
            }
            if (kb.childOf(term, "Attribute")) {
                result.add(TermAttr(CHILDOFATTRIBUTE))
            }

            if (kb.isSubclass(term, "Relation")) {
                result.add(TermAttr(SUBCLASSOFRELATION))
            }
            if (kb.isSubclass(term, "Predicate")) {
                result.add(TermAttr(SUBCLASSOFPREDICATE))
            }
            if (kb.isSubclass(term, "Function")) {
                result.add(TermAttr(SUBCLASSOFFUNCTION))
            }
            if (kb.isSubclass(term, "Attribute")) {
                result.add(TermAttr(SUBCLASSOFATTRIBUTE))
            }

            if (result.isEmpty()) throw NotFoundException(term)

            return result
        }
    }
}
